{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "christian-celebration",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "import gzip\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import math\n",
    "from functools import reduce\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "threaded-morocco",
   "metadata": {},
   "source": [
    "# Selección de regiones del genoma de N. benthamiana en base a su perfil transcriptómico"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "normal-bulletin",
   "metadata": {},
   "source": [
    "### Step 1:\n",
    "Importamos un archivo con los identificadores de los genes de Niben261 con valores de expresión (gen correspondiente en Niben101), la posición inicial y final en el cromosoma, cromosoma y categoría de expresión a la que pertenecce y un archivo con todos los genes ano."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fifth-internet",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\milil\\anaconda3\\lib\\site-packages\\ipykernel_launcher.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>quintile</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Niben261Chr01g0002008</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>266317</td>\n",
       "      <td>268851</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Niben261Chr01g0004001</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>458230</td>\n",
       "      <td>462273</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Niben261Chr01g0004003</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>467574</td>\n",
       "      <td>480423</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Niben261Chr01g0004011</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>465674</td>\n",
       "      <td>466936</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Niben261Chr01g0004012</th>\n",
       "      <td>Chr01</td>\n",
       "      <td>432045</td>\n",
       "      <td>433043</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Niben261Chr19g1484002</th>\n",
       "      <td>Chr19</td>\n",
       "      <td>148454367</td>\n",
       "      <td>148463060</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Niben261Chr19g1485005</th>\n",
       "      <td>Chr19</td>\n",
       "      <td>148480135</td>\n",
       "      <td>148499120</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Niben261Chr19g1485014</th>\n",
       "      <td>Chr19</td>\n",
       "      <td>148530961</td>\n",
       "      <td>148538381</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Niben261Chr19g1486029</th>\n",
       "      <td>Chr19</td>\n",
       "      <td>148623104</td>\n",
       "      <td>148631519</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Niben261Chr19g1486004</th>\n",
       "      <td>Chr19</td>\n",
       "      <td>148632519</td>\n",
       "      <td>148634027</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60260 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         Chr      Start        End  quintile\n",
       "Niben261Chr01g0002008  Chr01     266317     268851       1.0\n",
       "Niben261Chr01g0004001  Chr01     458230     462273       3.0\n",
       "Niben261Chr01g0004003  Chr01     467574     480423       2.0\n",
       "Niben261Chr01g0004011  Chr01     465674     466936       1.0\n",
       "Niben261Chr01g0004012  Chr01     432045     433043       1.0\n",
       "...                      ...        ...        ...       ...\n",
       "Niben261Chr19g1484002  Chr19  148454367  148463060       NaN\n",
       "Niben261Chr19g1485005  Chr19  148480135  148499120       NaN\n",
       "Niben261Chr19g1485014  Chr19  148530961  148538381       NaN\n",
       "Niben261Chr19g1486029  Chr19  148623104  148631519       NaN\n",
       "Niben261Chr19g1486004  Chr19  148632519  148634027       NaN\n",
       "\n",
       "[60260 rows x 4 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.read_csv(\"./raw_data/data_quintile_round_inNiben261.txt\", sep = \"\\t\", header = 0)\n",
    "Niben261 = pd.read_csv(\"./raw_data/Niben261_genes.txt\", sep = \"\\t\", header = 0, index_col = 0)\n",
    "Niben261chr = Niben261[Niben261.index.str.contains(\"Chr\")]\n",
    "# Complete data with Niben261 genes.\n",
    "quintile_genes = data.index.tolist() # genes clasificados, con correspondencia en Niben101.\n",
    "add_genes = list(set(Niben261chr.index.tolist()).difference(quintile_genes))\n",
    "Niben261add = Niben261chr[Niben261chr.index.isin(add_genes)]\n",
    "Niben261add[\"Chr\"] = Niben261add.Chr.str.replace(\"Niben261Chr\", \"Chr\")\n",
    "data_complete = data.append(Niben261add)\n",
    "data_complete"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "awful-terminal",
   "metadata": {},
   "source": [
    "### Step 2:\n",
    "Longitud para cada cromosoma."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "strange-personal",
   "metadata": {},
   "outputs": [],
   "source": [
    "chrs = sorted(list(set(data_complete[\"Chr\"]))) # List of chromosomes."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "distributed-composition",
   "metadata": {},
   "source": [
    "El dict se obtiene ejecutando esto, pero tarda mucho:\n",
    "\n",
    "Niben261fasta = \"C:/Users/milil/Downloads/Nextcloud/Documents/Miriam/Javi/Expression/Database/Genome/v261/Niben261_genome.fasta.gz\"\n",
    "chr_len = {} # Dict with chromosomes length.\n",
    "with gzip.open(Niben261fasta, \"rt\") as handle:\n",
    "    for record in SeqIO.parse(handle, \"fasta\"):\n",
    "        chr_len[str(record.id)] = len(str(record.seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "velvet-portfolio",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr_len = {'Niben261Chr01': 186295347,\n",
    " 'Niben261Chr02': 151625522,\n",
    " 'Niben261Chr03': 194605305,\n",
    " 'Niben261Chr04': 189815799,\n",
    " 'Niben261Chr05': 138451958,\n",
    " 'Niben261Chr06': 148966868,\n",
    " 'Niben261Chr07': 152648067,\n",
    " 'Niben261Chr08': 137820123,\n",
    " 'Niben261Chr09': 137640743,\n",
    " 'Niben261Chr10': 138251820,\n",
    " 'Niben261Chr11': 143669543,\n",
    " 'Niben261Chr12': 158218613,\n",
    " 'Niben261Chr13': 153652170,\n",
    " 'Niben261Chr14': 153778118,\n",
    " 'Niben261Chr15': 153112746,\n",
    " 'Niben261Chr16': 148817306,\n",
    " 'Niben261Chr17': 155116228,\n",
    " 'Niben261Chr18': 148425545,\n",
    " 'Niben261Chr19': 148948562}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "resident-reporter",
   "metadata": {},
   "source": [
    "### Step 3:\n",
    "**Estudio de los cromosomas para selección de ventana**\n",
    "\n",
    "Conclusiones sobre las interspace zones:\n",
    "* La ventana para estudiar la idoneidad de una región en base a los datos de expresión tendrá que abarcar por lo menos varios de ellos.\n",
    "* La media de las interspace zones más grandes es de 787,578.94.\n",
    "* Si tenemos una media de 44,701.88 bp en los cromosomas, la ventana tendrá que ser de varios cientos de miles de bp.\n",
    "\n",
    "**Selección de parámetros**\n",
    "* Ventana de 1000K.\n",
    "* Coef [4, 2 , 1, -1, -2, -4].\n",
    "* Eliminación de zonas sin genes.\n",
    "* Al menos una región intergénica de 10K: lo cumplen todas.\n",
    "\n",
    "*Los detalles están en score.ipynb*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "material-vocabulary",
   "metadata": {},
   "source": [
    "### Step 4:\n",
    "Cálculo de la idoneidad en el genoma."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "assured-creation",
   "metadata": {},
   "outputs": [],
   "source": [
    "def split(length, window):\n",
    "    intervals = {}\n",
    "    index = 1\n",
    "    interval = []\n",
    "    number_intervals = math.ceil(length/window)\n",
    "    for i in range(length):\n",
    "        interval.append(i + 1)\n",
    "        if index == number_intervals:\n",
    "            intervals[index] = [intervals[index - 1][1] + 1, length]\n",
    "            break\n",
    "        if len(interval) == window:\n",
    "            intervals[index] = [min(interval), max(interval)]\n",
    "            index += 1\n",
    "            interval = []\n",
    "    return(intervals)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "daily-guard",
   "metadata": {},
   "outputs": [],
   "source": [
    "coef = {\"x1\": 4,\n",
    "        \"x2\": 2,\n",
    "        \"x3\": 1,\n",
    "        \"x4\": -1,\n",
    "        \"x5\": -2,\n",
    "        \"x6\": -4}\n",
    "\n",
    "window = 1000000\n",
    "\n",
    "class Chr():\n",
    "    def __init__(self, chr, length, window, data_complete, coef):\n",
    "        self.chr = chr\n",
    "        self.length = length\n",
    "        self.window = window\n",
    "        self.data_complete = data_complete\n",
    "        self.coef = coef\n",
    "        self.intervals_chr = split(length, window)\n",
    "        self.idoneidad = self.idon()\n",
    "        \n",
    "    def idon(self):\n",
    "        datachr = self.data_complete[self.data_complete[\"Chr\"] == self.chr].sort_values('Start')\n",
    "        \n",
    "        self.idoneidad = {}\n",
    "        for interval, values in self.intervals_chr.items():\n",
    "            chr_values = datachr[(values[0] <= datachr[\"Start\"]) & (datachr[\"Start\"] < values[1])]\n",
    "            n_genes = len(chr_values[chr_values['quintile'].notna()])\n",
    "            if n_genes != 0:\n",
    "                z = self.coef[\"x1\"] * len(chr_values[chr_values[\"quintile\"] == 5]) + self.coef[\"x2\"] * len(chr_values[chr_values[\"quintile\"] == 4]) + self.coef[\"x3\"] * len(chr_values[chr_values[\"quintile\"] == 3]) + self.coef[\"x4\"] * len(chr_values[chr_values[\"quintile\"] == 2]) + self.coef[\"x5\"] * len(chr_values[chr_values[\"quintile\"] == 1]) + self.coef[\"x6\"] * len(chr_values[chr_values[\"quintile\"] == 0])\n",
    "                self.idoneidad[str(values[0]) + \"-\" + str(values[1])] = z\n",
    "        return(self.idoneidad)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "regional-necessity",
   "metadata": {},
   "outputs": [],
   "source": [
    "idoneidad_dict = {}\n",
    "for chr in chrs:\n",
    "    print(chr)\n",
    "    idoneidad_dict[chr] = Chr(chr, chr_len[\"Niben261\" + chr], window, data_complete, coef)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "accepted-shift",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_temp_list = []\n",
    "for k, v in idoneidad_dict.items():\n",
    "    #print(v.idoneidad)\n",
    "    data_temp = pd.DataFrame(v.idoneidad.items(), columns=['Interval', 'Suitability'])\n",
    "    data_temp[\"Chr\"] = k\n",
    "    data_temp[\"Window\"] = range(1, len(v.idoneidad.keys()) + 1, 1)\n",
    "    data_temp_list.append(data_temp)\n",
    "\n",
    "data_idoneidad = pd.concat(data_temp_list)\n",
    "data_idoneidad[[\"Start\", \"End\"]] = data_idoneidad[\"Interval\"].str.split('-', 1, expand=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "expanded-tiffany",
   "metadata": {},
   "outputs": [],
   "source": [
    "#data_idoneidad.to_csv(\"./results/idoneidad_v4.txt\", sep = \"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "integrated-electronics",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_idoneidad = pd.read_csv(\"./results/idoneidad_v4.txt\", sep = \"\\t\") # Para no ejecutar la función que tarda mucho."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fiscal-atlantic",
   "metadata": {},
   "source": [
    "### Step 5:\n",
    "Estudio de características de las regiones intergénicas y de la ventana para seleccionar mejor las ventanas, no solo en base a su idoneidad/Suitability.\n",
    "\n",
    "Características:\n",
    "* Longitud media de las IR de la ventana.\n",
    "* Número de genes asignados a una categoría de expresión en la ventana.\n",
    "* Número de genes totales en la ventana.\n",
    "* % GC medio de las IR de la ventana.\n",
    "* % N medio de las IR de la ventana.\n",
    "* Número de IR de la ventana."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cognitive-escape",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Chrs seqs.\n",
    "\n",
    "Niben261fasta = \"C:/Users/milil/Downloads/Nextcloud/Documents/Miriam/Javi/Expression/Database/Genome/v261/Niben261_genome.fasta.gz\"\n",
    "chr_seq = {} # Dict with chromosomes seq.\n",
    "with gzip.open(Niben261fasta, \"rt\") as handle:\n",
    "    for record in SeqIO.parse(handle, \"fasta\"):\n",
    "        chr_seq[str(record.id)] = str(record.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "exotic-oregon",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Chr_length():\n",
    "    def __init__(self, chr, data_idoneidad, data_complete, chr_seq):\n",
    "        self.chr = chr\n",
    "        self.data_idoneidad = data_idoneidad\n",
    "        self.data_complete = data_complete\n",
    "        self.chr_seq = chr_seq\n",
    "        self.length_intergenic, self.requisite, self.n_genes_quintile, self.n_genes_total, self.intergenic_regions, self.GC, self.N, self.n_intergenic_regions = self.idon_length()\n",
    "        \n",
    "    def idon_length(self):\n",
    "        datachr = self.data_complete[self.data_complete[\"Chr\"] == self.chr].sort_values('Start')\n",
    "        data_idoneidadchr = self.data_idoneidad[self.data_idoneidad[\"Chr\"] == self.chr]\n",
    "        \n",
    "        self.length_intergenic = {}\n",
    "        self.requisite = {}\n",
    "        self.n_genes_quintile = {}\n",
    "        self.n_genes_total = {}\n",
    "        self.intergenic_regions = {}\n",
    "        self.GC = {}\n",
    "        self.N = {}\n",
    "        self.n_intergenic_regions = {}\n",
    "        for index, row in data_idoneidadchr.iterrows():\n",
    "            self.seq = {}\n",
    "            start = int(row[\"Interval\"].split(\"-\")[0])\n",
    "            end = int(row[\"Interval\"].split(\"-\")[1])\n",
    "            list_intergenic_length = []\n",
    "            list_intergenic_GC = []\n",
    "            list_intergenic_N = []\n",
    "            requisite = 0 # Si tienen un región intergénica de al menos 10K será igual a 1.\n",
    "\n",
    "            chr_subset = datachr[(start <= datachr[\"Start\"]) & (datachr[\"Start\"] < end)].sort_values(\"Start\")\n",
    "            if len(chr_subset) == 0:\n",
    "                #list_intergenic_length.append(end-start + 1) # Se suman los extremos.\n",
    "                # Si no hay genes en la ventana no se considera.\n",
    "                pass\n",
    "\n",
    "            else:\n",
    "                count = 0\n",
    "                n_IR = 0\n",
    "                for i, r in chr_subset.iterrows():\n",
    "                    if count == 0:\n",
    "                        if r[\"Start\"] > start:\n",
    "                            intergenic_length = r[\"Start\"] - start\n",
    "                            list_intergenic_length.append(intergenic_length)\n",
    "                            intergenic_seq = self.chr_seq[\"Niben261\" + self.chr][start-1:r[\"Start\"]-1].upper()\n",
    "                            n_IR += 1\n",
    "                            self.seq[\"IR_\" + str(n_IR) + \";\" + str(start) + \"-\" + str(r[\"Start\"]-1)] = intergenic_seq.upper()\n",
    "                            GC_content = ((intergenic_seq.count('G')+intergenic_seq.count('C'))/len(intergenic_seq))*100\n",
    "                            list_intergenic_GC.append(GC_content)\n",
    "                            N_content = (intergenic_seq.count('N')/len(intergenic_seq))*100\n",
    "                            list_intergenic_N.append(N_content)\n",
    "                            #intergenic_seq_GC = intergenic_seq # [x:y] la y se cuenta y-1, así que se deja x-1 e y porque la seq empieza por 0 y no por 1.\n",
    "                        end_new = r[\"End\"]\n",
    "                    else:\n",
    "                        intergenic_length = r[\"Start\"] - end_new - 1 # No coger los extremos. Sólo en el caso de las IR entre genes, en el caso de la IR del principio y del final no hace falta.\n",
    "                        if intergenic_length > 0:\n",
    "                            list_intergenic_length.append(intergenic_length)\n",
    "                            intergenic_seq = self.chr_seq[\"Niben261\" + self.chr][end_new:r[\"Start\"]-1].upper()\n",
    "                            GC_content = ((intergenic_seq.count('G')+intergenic_seq.count('C'))/len(intergenic_seq))*100\n",
    "                            list_intergenic_GC.append(GC_content)\n",
    "                            N_content = (intergenic_seq.count('N')/len(intergenic_seq))*100\n",
    "                            list_intergenic_N.append(N_content)\n",
    "                            #intergenic_seq_GC = intergenic_seq\n",
    "                            n_IR += 1\n",
    "                            self.seq[\"IR_\" + str(n_IR) + \";\" + str(end_new + 1) + \"-\" + str(r[\"Start\"]-1)] = intergenic_seq\n",
    "                        end_new = r[\"End\"]\n",
    "                    #list_intergenic_length.append(intergenic_length)\n",
    "                    count += 1\n",
    "                    #self.seq[\"IR_\" + str(index)] = intergenic_seq\n",
    "                    if count == len(chr_subset):\n",
    "                        if r[\"End\"] < end:\n",
    "                            intergenic_length = end - r[\"End\"] \n",
    "                            intergenic_seq = self.chr_seq[\"Niben261\" + self.chr][r[\"End\"]:end].upper()\n",
    "                            GC_content = ((intergenic_seq.count('G')+intergenic_seq.count('C'))/len(intergenic_seq))*100\n",
    "                            #intergenic_seq_GC = intergenic_seq\n",
    "                            list_intergenic_GC.append(GC_content)\n",
    "                            N_content = (intergenic_seq.count('N')/len(intergenic_seq))*100\n",
    "                            list_intergenic_N.append(N_content)\n",
    "                            list_intergenic_length.append(intergenic_length)\n",
    "                            n_IR += 1\n",
    "                            self.seq[\"IR_\" + str(n_IR) + \";\" + str(r[\"End\"]+1) + \"-\" + str(end)] = intergenic_seq\n",
    "                if any(i >= 10000 for i in list_intergenic_length): requisite = 1\n",
    "                #self.length_intergenic[row[\"Interval\"]] = sum(list_intergenic_length)\n",
    "                self.length_intergenic[row[\"Interval\"]] = np.mean(list_intergenic_length)\n",
    "                self.requisite[row[\"Interval\"]] = requisite\n",
    "                self.n_genes_quintile[row[\"Interval\"]] = len(chr_subset[chr_subset['quintile'].notna()]) # n of genes assigned to one expression category.\n",
    "                self.n_genes_total[row[\"Interval\"]] = len(chr_subset) # n of total genes.\n",
    "                self.intergenic_regions[row[\"Interval\"]] = self.seq\n",
    "                self.GC[row[\"Interval\"]] = np.mean(list_intergenic_GC)\n",
    "                self.N[row[\"Interval\"]] = np.mean(list_intergenic_N)\n",
    "                self.n_intergenic_regions[row[\"Interval\"]] = n_IR\n",
    "        return(self.length_intergenic, self.requisite, self.n_genes_quintile, self.n_genes_total, self.intergenic_regions, self.GC, self.N, self.n_intergenic_regions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "restricted-massachusetts",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chr01\n",
      "Chr02\n",
      "Chr03\n",
      "Chr04\n",
      "Chr05\n",
      "Chr06\n",
      "Chr07\n",
      "Chr08\n",
      "Chr09\n",
      "Chr10\n",
      "Chr11\n",
      "Chr12\n",
      "Chr13\n",
      "Chr14\n",
      "Chr15\n",
      "Chr16\n",
      "Chr17\n",
      "Chr18\n",
      "Chr19\n"
     ]
    }
   ],
   "source": [
    "idoneidad_intergenic = {}\n",
    "for chr in chrs:\n",
    "    print(chr)\n",
    "    idoneidad_intergenic[chr] = Chr_length(chr, data_idoneidad, data_complete, chr_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "intense-attack",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Interval</th>\n",
       "      <th>Suitability</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Window</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Mean Length IR</th>\n",
       "      <th>requisite</th>\n",
       "      <th>N genes assigned to expression category</th>\n",
       "      <th>N genes total</th>\n",
       "      <th>Mean % GC</th>\n",
       "      <th>Mean % N</th>\n",
       "      <th>N IR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-1000000</td>\n",
       "      <td>-11</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1000000</td>\n",
       "      <td>30352.793103</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>29</td>\n",
       "      <td>33.534990</td>\n",
       "      <td>9.317688</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1000001-2000000</td>\n",
       "      <td>-26</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>2</td>\n",
       "      <td>1000001</td>\n",
       "      <td>2000000</td>\n",
       "      <td>24140.558824</td>\n",
       "      <td>1</td>\n",
       "      <td>28</td>\n",
       "      <td>35</td>\n",
       "      <td>32.043833</td>\n",
       "      <td>9.104434</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2000001-3000000</td>\n",
       "      <td>-15</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>3</td>\n",
       "      <td>2000001</td>\n",
       "      <td>3000000</td>\n",
       "      <td>37669.916667</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>23</td>\n",
       "      <td>34.765797</td>\n",
       "      <td>5.991842</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3000001-4000000</td>\n",
       "      <td>-30</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>4</td>\n",
       "      <td>3000001</td>\n",
       "      <td>4000000</td>\n",
       "      <td>25096.171429</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>34</td>\n",
       "      <td>34.479037</td>\n",
       "      <td>8.040428</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4000001-5000000</td>\n",
       "      <td>-7</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>5</td>\n",
       "      <td>4000001</td>\n",
       "      <td>5000000</td>\n",
       "      <td>73944.153846</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>13</td>\n",
       "      <td>33.628559</td>\n",
       "      <td>9.327155</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2935</th>\n",
       "      <td>144000001-145000000</td>\n",
       "      <td>-29</td>\n",
       "      <td>Chr19</td>\n",
       "      <td>145</td>\n",
       "      <td>144000001</td>\n",
       "      <td>145000000</td>\n",
       "      <td>37572.166667</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "      <td>23</td>\n",
       "      <td>33.537813</td>\n",
       "      <td>10.177448</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2936</th>\n",
       "      <td>145000001-146000000</td>\n",
       "      <td>-37</td>\n",
       "      <td>Chr19</td>\n",
       "      <td>146</td>\n",
       "      <td>145000001</td>\n",
       "      <td>146000000</td>\n",
       "      <td>17201.872340</td>\n",
       "      <td>1</td>\n",
       "      <td>33</td>\n",
       "      <td>46</td>\n",
       "      <td>34.315892</td>\n",
       "      <td>6.056474</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2937</th>\n",
       "      <td>146000001-147000000</td>\n",
       "      <td>-14</td>\n",
       "      <td>Chr19</td>\n",
       "      <td>147</td>\n",
       "      <td>146000001</td>\n",
       "      <td>147000000</td>\n",
       "      <td>31505.962963</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>26</td>\n",
       "      <td>32.106336</td>\n",
       "      <td>12.715414</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2938</th>\n",
       "      <td>147000001-148000000</td>\n",
       "      <td>-14</td>\n",
       "      <td>Chr19</td>\n",
       "      <td>148</td>\n",
       "      <td>147000001</td>\n",
       "      <td>148000000</td>\n",
       "      <td>41179.409091</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "      <td>22</td>\n",
       "      <td>32.165613</td>\n",
       "      <td>11.726650</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2939</th>\n",
       "      <td>148000001-148948562</td>\n",
       "      <td>-34</td>\n",
       "      <td>Chr19</td>\n",
       "      <td>149</td>\n",
       "      <td>148000001</td>\n",
       "      <td>148948562</td>\n",
       "      <td>35824.086957</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>23</td>\n",
       "      <td>34.768534</td>\n",
       "      <td>7.512156</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2940 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Interval  Suitability    Chr  Window      Start        End  \\\n",
       "0               1-1000000          -11  Chr01       1          1    1000000   \n",
       "1         1000001-2000000          -26  Chr01       2    1000001    2000000   \n",
       "2         2000001-3000000          -15  Chr01       3    2000001    3000000   \n",
       "3         3000001-4000000          -30  Chr01       4    3000001    4000000   \n",
       "4         4000001-5000000           -7  Chr01       5    4000001    5000000   \n",
       "...                   ...          ...    ...     ...        ...        ...   \n",
       "2935  144000001-145000000          -29  Chr19     145  144000001  145000000   \n",
       "2936  145000001-146000000          -37  Chr19     146  145000001  146000000   \n",
       "2937  146000001-147000000          -14  Chr19     147  146000001  147000000   \n",
       "2938  147000001-148000000          -14  Chr19     148  147000001  148000000   \n",
       "2939  148000001-148948562          -34  Chr19     149  148000001  148948562   \n",
       "\n",
       "      Mean Length IR  requisite  N genes assigned to expression category  \\\n",
       "0       30352.793103          1                                       15   \n",
       "1       24140.558824          1                                       28   \n",
       "2       37669.916667          1                                        9   \n",
       "3       25096.171429          1                                       25   \n",
       "4       73944.153846          1                                        9   \n",
       "...              ...        ...                                      ...   \n",
       "2935    37572.166667          1                                       17   \n",
       "2936    17201.872340          1                                       33   \n",
       "2937    31505.962963          1                                       19   \n",
       "2938    41179.409091          1                                       17   \n",
       "2939    35824.086957          1                                       11   \n",
       "\n",
       "      N genes total  Mean % GC   Mean % N  N IR  \n",
       "0                29  33.534990   9.317688    29  \n",
       "1                35  32.043833   9.104434    34  \n",
       "2                23  34.765797   5.991842    24  \n",
       "3                34  34.479037   8.040428    35  \n",
       "4                13  33.628559   9.327155    13  \n",
       "...             ...        ...        ...   ...  \n",
       "2935             23  33.537813  10.177448    24  \n",
       "2936             46  34.315892   6.056474    47  \n",
       "2937             26  32.106336  12.715414    27  \n",
       "2938             22  32.165613  11.726650    22  \n",
       "2939             23  34.768534   7.512156    23  \n",
       "\n",
       "[2940 rows x 13 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Table 1:\n",
    "\n",
    "data_length_temp_list = []\n",
    "for k, v in idoneidad_intergenic.items():\n",
    "    #print(v.idoneidad)\n",
    "    data_temp_intergenic = pd.DataFrame(v.length_intergenic.items(), columns=['Interval', 'Mean Length IR'])\n",
    "    data_temp_requisite = pd.DataFrame(v.requisite.items(), columns=['Interval', 'requisite'])\n",
    "    data_temp_n_genes_quintile = pd.DataFrame(v.n_genes_quintile.items(), columns=['Interval', 'N genes assigned to expression category'])\n",
    "    data_temp_n_genes_total = pd.DataFrame(v.n_genes_total.items(), columns=['Interval', 'N genes total'])\n",
    "    data_temp_GC = pd.DataFrame(v.GC.items(), columns=['Interval', 'Mean % GC'])\n",
    "    data_temp_N = pd.DataFrame(v.N.items(), columns=['Interval', 'Mean % N'])\n",
    "    data_temp_n_intergenic_regions = pd.DataFrame(v.n_intergenic_regions.items(), columns=['Interval', 'N IR'])\n",
    "    data_temp_intergenic[\"Chr\"] = k\n",
    "    dfs = [data_temp_intergenic, data_temp_requisite, data_temp_n_genes_quintile, data_temp_n_genes_total, data_temp_GC, data_temp_N, data_temp_n_intergenic_regions]\n",
    "    df_temp = reduce(lambda left,right: pd.merge(left,right,on='Interval'), dfs)\n",
    "    data_length_temp_list.append(df_temp)\n",
    "\n",
    "data_intergenic = pd.concat(data_length_temp_list)\n",
    "data_idoneidad_intergenic = pd.merge(data_idoneidad, data_intergenic, on=['Interval', 'Chr'], how='outer')\n",
    "data_idoneidad_intergenic = data_idoneidad_intergenic[data_idoneidad_intergenic.columns[~data_idoneidad_intergenic.columns.str.contains('Unnamed:')]]\n",
    "data_idoneidad_intergenic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "international-comfort",
   "metadata": {},
   "outputs": [],
   "source": [
    "#data_idoneidad_intergenic.to_csv(\"./results/findIR_complete_v2.txt\", sep = \"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "exciting-response",
   "metadata": {},
   "outputs": [],
   "source": [
    "# File with all the data.\n",
    "data_idoneidad_intergenic = pd.read_csv(\"./results/findIR_complete_v2.txt\", sep = \"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "experimental-candy",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Interval</th>\n",
       "      <th>Suitability</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Window</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Mean Length IR</th>\n",
       "      <th>requisite</th>\n",
       "      <th>N genes assigned to expression category</th>\n",
       "      <th>N genes total</th>\n",
       "      <th>Mean % GC</th>\n",
       "      <th>Mean % N</th>\n",
       "      <th>N IR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1895</th>\n",
       "      <td>1895</td>\n",
       "      <td>17000001-18000000</td>\n",
       "      <td>9</td>\n",
       "      <td>Chr13</td>\n",
       "      <td>18</td>\n",
       "      <td>17000001</td>\n",
       "      <td>18000000</td>\n",
       "      <td>41276.666667</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>20</td>\n",
       "      <td>35.146000</td>\n",
       "      <td>5.736527</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1584</th>\n",
       "      <td>1584</td>\n",
       "      <td>9000001-10000000</td>\n",
       "      <td>6</td>\n",
       "      <td>Chr11</td>\n",
       "      <td>10</td>\n",
       "      <td>9000001</td>\n",
       "      <td>10000000</td>\n",
       "      <td>73755.461538</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>13</td>\n",
       "      <td>36.228817</td>\n",
       "      <td>7.296826</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1426</th>\n",
       "      <td>1426</td>\n",
       "      <td>127000001-128000000</td>\n",
       "      <td>6</td>\n",
       "      <td>Chr09</td>\n",
       "      <td>128</td>\n",
       "      <td>127000001</td>\n",
       "      <td>128000000</td>\n",
       "      <td>36060.360000</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>24</td>\n",
       "      <td>34.396043</td>\n",
       "      <td>5.753853</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1993</th>\n",
       "      <td>1993</td>\n",
       "      <td>115000001-116000000</td>\n",
       "      <td>6</td>\n",
       "      <td>Chr13</td>\n",
       "      <td>116</td>\n",
       "      <td>115000001</td>\n",
       "      <td>116000000</td>\n",
       "      <td>36698.080000</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>24</td>\n",
       "      <td>33.734135</td>\n",
       "      <td>5.430588</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2157</th>\n",
       "      <td>2157</td>\n",
       "      <td>125000001-126000000</td>\n",
       "      <td>6</td>\n",
       "      <td>Chr14</td>\n",
       "      <td>126</td>\n",
       "      <td>125000001</td>\n",
       "      <td>126000000</td>\n",
       "      <td>71668.230769</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>12</td>\n",
       "      <td>33.145199</td>\n",
       "      <td>9.121484</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1867</th>\n",
       "      <td>1867</td>\n",
       "      <td>148000001-149000000</td>\n",
       "      <td>-64</td>\n",
       "      <td>Chr12</td>\n",
       "      <td>149</td>\n",
       "      <td>148000001</td>\n",
       "      <td>149000000</td>\n",
       "      <td>20683.590909</td>\n",
       "      <td>1</td>\n",
       "      <td>27</td>\n",
       "      <td>43</td>\n",
       "      <td>40.883291</td>\n",
       "      <td>2.197469</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2496</th>\n",
       "      <td>2496</td>\n",
       "      <td>9000001-10000000</td>\n",
       "      <td>-66</td>\n",
       "      <td>Chr17</td>\n",
       "      <td>10</td>\n",
       "      <td>9000001</td>\n",
       "      <td>10000000</td>\n",
       "      <td>32603.750000</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>27</td>\n",
       "      <td>32.587732</td>\n",
       "      <td>10.553918</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2027</th>\n",
       "      <td>2027</td>\n",
       "      <td>149000001-150000000</td>\n",
       "      <td>-66</td>\n",
       "      <td>Chr13</td>\n",
       "      <td>150</td>\n",
       "      <td>149000001</td>\n",
       "      <td>150000000</td>\n",
       "      <td>25910.166667</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>37</td>\n",
       "      <td>42.441372</td>\n",
       "      <td>0.115343</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2026</th>\n",
       "      <td>2026</td>\n",
       "      <td>148000001-149000000</td>\n",
       "      <td>-96</td>\n",
       "      <td>Chr13</td>\n",
       "      <td>149</td>\n",
       "      <td>148000001</td>\n",
       "      <td>149000000</td>\n",
       "      <td>23360.475000</td>\n",
       "      <td>1</td>\n",
       "      <td>26</td>\n",
       "      <td>41</td>\n",
       "      <td>43.659821</td>\n",
       "      <td>0.005470</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2028</th>\n",
       "      <td>2028</td>\n",
       "      <td>150000001-151000000</td>\n",
       "      <td>-133</td>\n",
       "      <td>Chr13</td>\n",
       "      <td>151</td>\n",
       "      <td>150000001</td>\n",
       "      <td>151000000</td>\n",
       "      <td>13156.323529</td>\n",
       "      <td>1</td>\n",
       "      <td>40</td>\n",
       "      <td>68</td>\n",
       "      <td>43.222241</td>\n",
       "      <td>0.251912</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2940 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unnamed: 0             Interval  Suitability    Chr  Window      Start  \\\n",
       "1895        1895    17000001-18000000            9  Chr13      18   17000001   \n",
       "1584        1584     9000001-10000000            6  Chr11      10    9000001   \n",
       "1426        1426  127000001-128000000            6  Chr09     128  127000001   \n",
       "1993        1993  115000001-116000000            6  Chr13     116  115000001   \n",
       "2157        2157  125000001-126000000            6  Chr14     126  125000001   \n",
       "...          ...                  ...          ...    ...     ...        ...   \n",
       "1867        1867  148000001-149000000          -64  Chr12     149  148000001   \n",
       "2496        2496     9000001-10000000          -66  Chr17      10    9000001   \n",
       "2027        2027  149000001-150000000          -66  Chr13     150  149000001   \n",
       "2026        2026  148000001-149000000          -96  Chr13     149  148000001   \n",
       "2028        2028  150000001-151000000         -133  Chr13     151  150000001   \n",
       "\n",
       "            End  Mean Length IR  requisite  \\\n",
       "1895   18000000    41276.666667          1   \n",
       "1584   10000000    73755.461538          1   \n",
       "1426  128000000    36060.360000          1   \n",
       "1993  116000000    36698.080000          1   \n",
       "2157  126000000    71668.230769          1   \n",
       "...         ...             ...        ...   \n",
       "1867  149000000    20683.590909          1   \n",
       "2496   10000000    32603.750000          1   \n",
       "2027  150000000    25910.166667          1   \n",
       "2026  149000000    23360.475000          1   \n",
       "2028  151000000    13156.323529          1   \n",
       "\n",
       "      N genes assigned to expression category  N genes total  Mean % GC  \\\n",
       "1895                                       14             20  35.146000   \n",
       "1584                                        6             13  36.228817   \n",
       "1426                                       16             24  34.396043   \n",
       "1993                                       12             24  33.734135   \n",
       "2157                                       10             12  33.145199   \n",
       "...                                       ...            ...        ...   \n",
       "1867                                       27             43  40.883291   \n",
       "2496                                       23             27  32.587732   \n",
       "2027                                       20             37  42.441372   \n",
       "2026                                       26             41  43.659821   \n",
       "2028                                       40             68  43.222241   \n",
       "\n",
       "       Mean % N  N IR  \n",
       "1895   5.736527    21  \n",
       "1584   7.296826    13  \n",
       "1426   5.753853    25  \n",
       "1993   5.430588    25  \n",
       "2157   9.121484    13  \n",
       "...         ...   ...  \n",
       "1867   2.197469    44  \n",
       "2496  10.553918    28  \n",
       "2027   0.115343    36  \n",
       "2026   0.005470    40  \n",
       "2028   0.251912    68  \n",
       "\n",
       "[2940 rows x 14 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_idoneidad_intergenic.sort_values('Suitability', ascending = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "mounted-marble",
   "metadata": {},
   "source": [
    "### Step 6:\n",
    "Generar fasta files con las secuencias de las regiones intergénicas por cromosoma y ventana.\n",
    "\n",
    "Ellos podrán calcular el % GC, la longitud, el % N y lo que quieran desde el fasta."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "monthly-suggestion",
   "metadata": {},
   "outputs": [],
   "source": [
    "for k, v in idoneidad_intergenic.items():\n",
    "    fastaFile = open(\"./IR_seqs/\" + k + \"_IR.fasta\", \"w+\")\n",
    "    for window, IRs in v.intergenic_regions.items():\n",
    "        for IR_id, IR_seq in IRs.items():\n",
    "            fastaFile.write(\">\" + k + \";\" + window + \";\" + IR_id + \"\\n\")\n",
    "            fastaFile.write(IR_seq.upper() + \"\\n\")\n",
    "    fastaFile.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "baking-contemporary",
   "metadata": {},
   "source": [
    "### Comprobaciones: para borrar después."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "higher-employer",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Selección de varias ventanas para hacer las comprobaciones: la mejor (mayor suitability), la peor, dos intermedias, una en el principio de un cromosoma, otra en el final de otro cromosoma.\n",
    "comprobaciones = {\n",
    "    \"Chr13_17000001-18000000\": (\"Chr13\", 17000001, 18000000), # Mejor ventana.\n",
    "    \"Chr13_150000001-151000000\": (\"Chr13\", 150000001, 151000000), # Peor ventana.\n",
    "    \"Chr02_142000001-143000000\": (\"Chr02\", 142000001, 143000000), # Ventana intermedia.\n",
    "    \"Chr10_99000001-100000000\": (\"Chr10\", 99000001, 100000000), # Ventana intermedia.\n",
    "    \"Chr07_1-1000000\": (\"Chr07\", 1, 1000000), # Primera ventana de un cromosoma.\n",
    "    \"Chr05_138000001-138451958\": (\"Chr05\", 138000001, 138451958), # Suitability = 0. Última ventana de un cromosoma.\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "disturbed-perception",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Interval</th>\n",
       "      <th>Suitability</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Window</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Mean Length IR</th>\n",
       "      <th>requisite</th>\n",
       "      <th>N genes assigned to expression category</th>\n",
       "      <th>N genes total</th>\n",
       "      <th>Mean % GC</th>\n",
       "      <th>Mean % N</th>\n",
       "      <th>N IR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1895</th>\n",
       "      <td>1895</td>\n",
       "      <td>17000001-18000000</td>\n",
       "      <td>9</td>\n",
       "      <td>Chr13</td>\n",
       "      <td>18</td>\n",
       "      <td>17000001</td>\n",
       "      <td>18000000</td>\n",
       "      <td>41276.666667</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>20</td>\n",
       "      <td>35.146000</td>\n",
       "      <td>5.736527</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2028</th>\n",
       "      <td>2028</td>\n",
       "      <td>150000001-151000000</td>\n",
       "      <td>-133</td>\n",
       "      <td>Chr13</td>\n",
       "      <td>151</td>\n",
       "      <td>150000001</td>\n",
       "      <td>151000000</td>\n",
       "      <td>13156.323529</td>\n",
       "      <td>1</td>\n",
       "      <td>40</td>\n",
       "      <td>68</td>\n",
       "      <td>43.222241</td>\n",
       "      <td>0.251912</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>329</th>\n",
       "      <td>329</td>\n",
       "      <td>142000001-143000000</td>\n",
       "      <td>-21</td>\n",
       "      <td>Chr02</td>\n",
       "      <td>143</td>\n",
       "      <td>142000001</td>\n",
       "      <td>143000000</td>\n",
       "      <td>64738.800000</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>14</td>\n",
       "      <td>32.489838</td>\n",
       "      <td>5.477867</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1536</th>\n",
       "      <td>1536</td>\n",
       "      <td>99000001-100000000</td>\n",
       "      <td>-4</td>\n",
       "      <td>Chr10</td>\n",
       "      <td>100</td>\n",
       "      <td>99000001</td>\n",
       "      <td>100000000</td>\n",
       "      <td>53170.388889</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>17</td>\n",
       "      <td>34.719107</td>\n",
       "      <td>3.299987</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1009</th>\n",
       "      <td>1009</td>\n",
       "      <td>1-1000000</td>\n",
       "      <td>-46</td>\n",
       "      <td>Chr07</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1000000</td>\n",
       "      <td>23214.324324</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>36</td>\n",
       "      <td>34.037078</td>\n",
       "      <td>8.076054</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>860</td>\n",
       "      <td>138000001-138451958</td>\n",
       "      <td>0</td>\n",
       "      <td>Chr05</td>\n",
       "      <td>139</td>\n",
       "      <td>138000001</td>\n",
       "      <td>138451958</td>\n",
       "      <td>86555.200000</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>29.966651</td>\n",
       "      <td>21.803010</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unnamed: 0             Interval  Suitability    Chr  Window      Start  \\\n",
       "1895        1895    17000001-18000000            9  Chr13      18   17000001   \n",
       "2028        2028  150000001-151000000         -133  Chr13     151  150000001   \n",
       "329          329  142000001-143000000          -21  Chr02     143  142000001   \n",
       "1536        1536   99000001-100000000           -4  Chr10     100   99000001   \n",
       "1009        1009            1-1000000          -46  Chr07       1          1   \n",
       "860          860  138000001-138451958            0  Chr05     139  138000001   \n",
       "\n",
       "            End  Mean Length IR  requisite  \\\n",
       "1895   18000000    41276.666667          1   \n",
       "2028  151000000    13156.323529          1   \n",
       "329   143000000    64738.800000          1   \n",
       "1536  100000000    53170.388889          1   \n",
       "1009    1000000    23214.324324          1   \n",
       "860   138451958    86555.200000          1   \n",
       "\n",
       "      N genes assigned to expression category  N genes total  Mean % GC  \\\n",
       "1895                                       14             20  35.146000   \n",
       "2028                                       40             68  43.222241   \n",
       "329                                         7             14  32.489838   \n",
       "1536                                        8             17  34.719107   \n",
       "1009                                       25             36  34.037078   \n",
       "860                                         4              4  29.966651   \n",
       "\n",
       "       Mean % N  N IR  \n",
       "1895   5.736527    21  \n",
       "2028   0.251912    68  \n",
       "329    5.477867    15  \n",
       "1536   3.299987    18  \n",
       "1009   8.076054    37  \n",
       "860   21.803010     5  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# En findIR_compelte.txt, estas ventanas tienen los siguientes valores:\n",
    "list_temp = []\n",
    "for k, v in comprobaciones.items():\n",
    "    chr_id = v[0]\n",
    "    start = v[1]\n",
    "    end = v[2]\n",
    "    chr_data_temp = data_idoneidad_intergenic[data_idoneidad_intergenic[\"Chr\"] == chr_id].sort_values('Start')\n",
    "    temp = chr_data_temp[(chr_data_temp[\"Start\"] >= start) & (chr_data_temp[\"Start\"] < end)]\n",
    "    list_temp.append(temp)\n",
    "data_filter = pd.concat(list_temp)\n",
    "data_filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "written-april",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "La ventana Chr13_17000001-18000000 tiene una z de 9\n",
      "La ventana Chr13_150000001-151000000 tiene una z de -133\n",
      "La ventana Chr02_142000001-143000000 tiene una z de -21\n",
      "La ventana Chr10_99000001-100000000 tiene una z de -4\n",
      "La ventana Chr07_1-1000000 tiene una z de -46\n",
      "La ventana Chr05_138000001-138451958 tiene una z de 0\n"
     ]
    }
   ],
   "source": [
    "# Comprobación del cálculo de suitability.\n",
    "\n",
    "coef = {\"x1\": 4, \"x2\":2, \"x3\":1, \"x4\":-1, \"x5\":-2, \"x6\":-4}\n",
    "\n",
    "for k, v in comprobaciones.items():\n",
    "    chr_id = v[0]\n",
    "    start = v[1]\n",
    "    end = v[2]\n",
    "    chr_data = data_complete[data_complete[\"Chr\"] == chr_id].sort_values('Start')\n",
    "    chr_values = chr_data[(chr_data[\"Start\"] >= start) & (chr_data[\"Start\"] < end)]\n",
    "    z = coef[\"x1\"] * len(chr_values[chr_values[\"quintile\"] == 5]) + coef[\"x2\"] * len(chr_values[chr_values[\"quintile\"] == 4]) + coef[\"x3\"] * len(chr_values[chr_values[\"quintile\"] == 3]) + coef[\"x4\"] * len(chr_values[chr_values[\"quintile\"] == 2]) + coef[\"x5\"] * len(chr_values[chr_values[\"quintile\"] == 1]) + coef[\"x6\"] * len(chr_values[chr_values[\"quintile\"] == 0])\n",
    "    print(\"La ventana \" + k + \" tiene una z de \" + str(z))\n",
    "                "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "painted-genealogy",
   "metadata": {},
   "source": [
    "**El cálculo de suitability es correcto para las zonas seleccionadas.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "leading-burst",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "La ventana Chr13_17000001-18000000 tiene una longitud de 1000000\n",
      "La ventana Chr13_150000001-151000000 tiene una longitud de 1000000\n",
      "La ventana Chr02_142000001-143000000 tiene una longitud de 1000000\n",
      "La ventana Chr10_99000001-100000000 tiene una longitud de 1000000\n",
      "La ventana Chr07_1-1000000 tiene una longitud de 1000000\n",
      "La ventana Chr05_138000001-138451958 tiene una longitud de 451958\n"
     ]
    }
   ],
   "source": [
    "# Comprobación de la longitud de ventana.\n",
    "\n",
    "for k, v in comprobaciones.items():\n",
    "    chr_id = v[0]\n",
    "    start = v[1]\n",
    "    end = v[2]\n",
    "    len_window = end - start + 1 # Se cuentan los extremos.\n",
    "    print(\"La ventana \" + k + \" tiene una longitud de \" + str(len_window))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "connected-jumping",
   "metadata": {},
   "source": [
    "**El cálculo de la longitud de ventana es correcto para las zonas seleccionadas.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "brutal-consumption",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "La ventana Chr13_17000001-18000000 tiene 14 asignados a categoría de expresión.\n",
      "La ventana Chr13_150000001-151000000 tiene 40 asignados a categoría de expresión.\n",
      "La ventana Chr02_142000001-143000000 tiene 7 asignados a categoría de expresión.\n",
      "La ventana Chr10_99000001-100000000 tiene 8 asignados a categoría de expresión.\n",
      "La ventana Chr07_1-1000000 tiene 25 asignados a categoría de expresión.\n",
      "La ventana Chr05_138000001-138451958 tiene 4 asignados a categoría de expresión.\n"
     ]
    }
   ],
   "source": [
    "# Comprobación del número de genes asignados a una categoría de expresión.\n",
    "\n",
    "for k, v in comprobaciones.items():\n",
    "    chr_id = v[0]\n",
    "    start = v[1]\n",
    "    end = v[2]\n",
    "    chr_data = data_complete[data_complete[\"Chr\"] == chr_id].sort_values('Start')\n",
    "    chr_values = chr_data[(chr_data[\"Start\"] >= start) & (chr_data[\"Start\"] < end)]\n",
    "    n_genes = len(chr_values[chr_values[\"quintile\"].notna()])\n",
    "    print(\"La ventana \" + k + \" tiene \" + str(n_genes) + \" asignados a categoría de expresión.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "understanding-plain",
   "metadata": {},
   "source": [
    "**El cálculo del número de genes asignados a una categoría de expresión es correcto para las zonas seleccionadas.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "positive-transport",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "La ventana Chr13_17000001-18000000 tiene 20 totales.\n",
      "La ventana Chr13_150000001-151000000 tiene 68 totales.\n",
      "La ventana Chr02_142000001-143000000 tiene 14 totales.\n",
      "La ventana Chr10_99000001-100000000 tiene 17 totales.\n",
      "La ventana Chr07_1-1000000 tiene 36 totales.\n",
      "La ventana Chr05_138000001-138451958 tiene 4 totales.\n"
     ]
    }
   ],
   "source": [
    "# Comprobación del número de genes totales.\n",
    "\n",
    "for k, v in comprobaciones.items():\n",
    "    chr_id = v[0]\n",
    "    start = v[1]\n",
    "    end = v[2]\n",
    "    chr_data = data_complete[data_complete[\"Chr\"] == chr_id].sort_values('Start')\n",
    "    chr_values = chr_data[(chr_data[\"Start\"] >= start) & (chr_data[\"Start\"] < end)]\n",
    "    n_genes = len(chr_values)\n",
    "    print(\"La ventana \" + k + \" tiene \" + str(n_genes) + \" totales.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "emerging-beast",
   "metadata": {},
   "source": [
    "**El cálculo del número de genes totales es correcto para las zonas seleccionadas.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "trained-valuation",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "La ventana Chr13_17000001-18000000 tiene 21 IR.\n",
      "La ventana Chr13_150000001-151000000 tiene 68 IR.\n",
      "La ventana Chr02_142000001-143000000 tiene 15 IR.\n",
      "La ventana Chr10_99000001-100000000 tiene 18 IR.\n",
      "La ventana Chr07_1-1000000 tiene 37 IR.\n",
      "La ventana Chr05_138000001-138451958 tiene 5 IR.\n"
     ]
    }
   ],
   "source": [
    "# Comprobación de número de IR.\n",
    "\n",
    "for k, v in comprobaciones.items():\n",
    "    chr_id = v[0]\n",
    "    start = v[1]\n",
    "    end = v[2]\n",
    "    chr_data = data_complete[data_complete[\"Chr\"] == chr_id].sort_values('Start')\n",
    "    chr_values = chr_data[(chr_data[\"Start\"] >= start) & (chr_data[\"Start\"] < end)]\n",
    "    \n",
    "    n_IR = 0\n",
    "    count = 0\n",
    "    for index, row in chr_values.iterrows():\n",
    "        if count == 0:\n",
    "            if row[\"Start\"] <= start: pass\n",
    "            else:\n",
    "                n_IR += 1\n",
    "            new_end = row[\"End\"]\n",
    "        else:\n",
    "            if row[\"Start\"] > new_end + 1:\n",
    "                n_IR += 1\n",
    "            new_end = row[\"End\"]\n",
    "        count += 1\n",
    "        if count == len(chr_values):\n",
    "            if row[\"End\"] < end: n_IR += 1\n",
    "    print(\"La ventana \" + k + \" tiene \" + str(n_IR) + \" IR.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "hungarian-index",
   "metadata": {},
   "source": [
    "**El cálculo del número de regiones intergénicas es correcto para las ventanas seleccionadas.**\n",
    "\n",
    "Comprobado en Excel y en los fasta también."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "expected-theme",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "La ventana Chr13_17000001-18000000 tiene 41276.666666666664 longitud media.\n",
      "La ventana Chr13_150000001-151000000 tiene 13156.323529411764 longitud media.\n",
      "La ventana Chr02_142000001-143000000 tiene 64738.8 longitud media.\n",
      "La ventana Chr10_99000001-100000000 tiene 53170.38888888889 longitud media.\n",
      "La ventana Chr07_1-1000000 tiene 23214.324324324323 longitud media.\n",
      "La ventana Chr05_138000001-138451958 tiene 86555.2 longitud media.\n"
     ]
    }
   ],
   "source": [
    "# Comprobación de la longitud media de IR.\n",
    "\n",
    "for k, v in comprobaciones.items():\n",
    "    chr_id = v[0]\n",
    "    start = v[1]\n",
    "    end = v[2]\n",
    "    chr_data = data_complete[data_complete[\"Chr\"] == chr_id].sort_values('Start')\n",
    "    chr_values = chr_data[(chr_data[\"Start\"] >= start) & (chr_data[\"Start\"] < end)]\n",
    "    \n",
    "    length_list = []\n",
    "    count = 0\n",
    "    for index, row in chr_values.iterrows():\n",
    "        if count == 0:\n",
    "            if row[\"Start\"] <= start: pass\n",
    "            else:\n",
    "                length = row[\"Start\"] - start\n",
    "                length_list.append(length)\n",
    "            new_end = row[\"End\"]\n",
    "        else:\n",
    "            if row[\"Start\"] > new_end + 1:\n",
    "                length = row[\"Start\"] - new_end - 1\n",
    "                length_list.append(length)\n",
    "            new_end = row[\"End\"]\n",
    "        count += 1\n",
    "        if count == len(chr_values):\n",
    "            if row[\"End\"] < end:\n",
    "                length = end - row[\"End\"]\n",
    "                length_list.append(length)\n",
    "    mean_length = np.mean(length_list)\n",
    "    print(\"La ventana \" + k + \" tiene \" + str(mean_length) + \" longitud media.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "gentle-convertible",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "La ventana Chr13_17000001-18000000 tiene 41276.666666666664 longitud media.\n",
      "La ventana Chr13_150000001-151000000 tiene 13156.323529411764 longitud media.\n",
      "La ventana Chr02_142000001-143000000 tiene 64738.8 longitud media.\n",
      "La ventana Chr10_99000001-100000000 tiene 53170.38888888889 longitud media.\n",
      "La ventana Chr07_1-1000000 tiene 23214.324324324323 longitud media.\n",
      "La ventana Chr05_138000001-138451958 tiene 86555.2 longitud media.\n"
     ]
    }
   ],
   "source": [
    "# Comprobación de la longitud media de IR 2.\n",
    "\n",
    "for k, v in comprobaciones.items():\n",
    "    chr_id = v[0]\n",
    "    start = v[1]\n",
    "    end = v[2]\n",
    "    chr_data = data_complete[data_complete[\"Chr\"] == chr_id].sort_values('Start')\n",
    "    chr_values = chr_data[(chr_data[\"Start\"] >= start) & (chr_data[\"Start\"] < end)]\n",
    "    seq = chr_seq[\"Niben261\" + chr_id]\n",
    "    \n",
    "    length_list = []\n",
    "    count = 0\n",
    "    for index, row in chr_values.iterrows():\n",
    "        if count == 0:\n",
    "            if row[\"Start\"] <= start: pass\n",
    "            else:\n",
    "                IR_seq = seq[start-1:row[\"Start\"]-1]\n",
    "                length = len(IR_seq)\n",
    "                length_list.append(length)\n",
    "            new_end = row[\"End\"]\n",
    "        else:\n",
    "            if row[\"Start\"] > new_end + 1:\n",
    "                IR_seq = seq[new_end:row[\"Start\"]-1]\n",
    "                length = len(IR_seq)\n",
    "                length_list.append(length)\n",
    "            new_end = row[\"End\"]\n",
    "        count += 1\n",
    "        if count == len(chr_values):\n",
    "            if row[\"End\"] < end:\n",
    "                IR_seq = seq[row[\"End\"]:end]\n",
    "                length = len(IR_seq)\n",
    "                length_list.append(length)\n",
    "    mean_length = np.mean(length_list)\n",
    "    print(\"La ventana \" + k + \" tiene \" + str(mean_length) + \" longitud media.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "analyzed-police",
   "metadata": {},
   "source": [
    "**El cálculo de la longitud media de las IR es correcto para las ventanas seleccionadas.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "satellite-service",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "La ventana Chr13_17000001-18000000 tiene 1 requisite.\n",
      "La ventana Chr13_150000001-151000000 tiene 1 requisite.\n",
      "La ventana Chr02_142000001-143000000 tiene 1 requisite.\n",
      "La ventana Chr10_99000001-100000000 tiene 1 requisite.\n",
      "La ventana Chr07_1-1000000 tiene 1 requisite.\n",
      "La ventana Chr05_138000001-138451958 tiene 1 requisite.\n"
     ]
    }
   ],
   "source": [
    "# Comprobación del requisito de las ventanas.\n",
    "\n",
    "for k, v in comprobaciones.items():\n",
    "    chr_id = v[0]\n",
    "    start = v[1]\n",
    "    end = v[2]\n",
    "    chr_data = data_complete[data_complete[\"Chr\"] == chr_id].sort_values('Start')\n",
    "    chr_values = chr_data[(chr_data[\"Start\"] >= start) & (chr_data[\"Start\"] < end)]\n",
    "    seq = chr_seq[\"Niben261\" + chr_id]\n",
    "    requisite = 0\n",
    "    \n",
    "    length_list = []\n",
    "    count = 0\n",
    "    for index, row in chr_values.iterrows():\n",
    "        if count == 0:\n",
    "            if row[\"Start\"] <= start: pass\n",
    "            else:\n",
    "                IR_seq = seq[start-1:row[\"Start\"]-1]\n",
    "                length = len(IR_seq)\n",
    "                length_list.append(length)\n",
    "            new_end = row[\"End\"]\n",
    "        else:\n",
    "            if row[\"Start\"] > new_end + 1:\n",
    "                IR_seq = seq[new_end:row[\"Start\"]-1]\n",
    "                length = len(IR_seq)\n",
    "                length_list.append(length)\n",
    "            new_end = row[\"End\"]\n",
    "        count += 1\n",
    "        if count == len(chr_values):\n",
    "            if row[\"End\"] < end:\n",
    "                IR_seq = seq[row[\"End\"]:end]\n",
    "                length = len(IR_seq)\n",
    "                length_list.append(length)\n",
    "    if any(item >= 10000 for item in length_list): requisite = 1\n",
    "    print(\"La ventana \" + k + \" tiene \" + str(requisite) + \" requisite.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "impressed-chemistry",
   "metadata": {},
   "source": [
    "**El cálculo del requisito es correcto para las ventanas seleccionadas.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "prostate-mortality",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "La ventana Chr13_17000001-18000000 tiene 35.14599982080838 % GC.\n",
      "La ventana Chr13_150000001-151000000 tiene 43.222241072760454 % GC.\n",
      "La ventana Chr02_142000001-143000000 tiene 32.48983816459409 % GC.\n",
      "La ventana Chr10_99000001-100000000 tiene 34.71910695483638 % GC.\n",
      "La ventana Chr07_1-1000000 tiene 34.03707807411949 % GC.\n",
      "La ventana Chr05_138000001-138451958 tiene 29.96665056144073 % GC.\n"
     ]
    }
   ],
   "source": [
    "# Comprobación del contenido medio de % GC.\n",
    "\n",
    "for k, v in comprobaciones.items():\n",
    "    chr_id = v[0]\n",
    "    start = v[1]\n",
    "    end = v[2]\n",
    "    chr_data = data_complete[data_complete[\"Chr\"] == chr_id].sort_values('Start')\n",
    "    chr_values = chr_data[(chr_data[\"Start\"] >= start) & (chr_data[\"Start\"] < end)]\n",
    "    seq = chr_seq[\"Niben261\" + chr_id]\n",
    "    \n",
    "    GC_list = []\n",
    "    count = 0\n",
    "    for index, row in chr_values.iterrows():\n",
    "        if count == 0:\n",
    "            if row[\"Start\"] <= start: pass\n",
    "            else:\n",
    "                IR_seq = seq[start-1:row[\"Start\"]-1].upper()\n",
    "                GC_content = ((IR_seq.count('G')+IR_seq.count('C'))/len(IR_seq))*100\n",
    "                GC_list.append(GC_content)\n",
    "            new_end = row[\"End\"]\n",
    "        else:\n",
    "            if row[\"Start\"] > new_end + 1:\n",
    "                IR_seq = seq[new_end:row[\"Start\"]-1].upper()\n",
    "                GC_content = ((IR_seq.count('G')+IR_seq.count('C'))/len(IR_seq))*100\n",
    "                GC_list.append(GC_content)\n",
    "            new_end = row[\"End\"]\n",
    "        count += 1\n",
    "        if count == len(chr_values):\n",
    "            if row[\"End\"] < end:\n",
    "                IR_seq = seq[row[\"End\"]:end].upper()\n",
    "                GC_content = ((IR_seq.count('G')+IR_seq.count('C'))/len(IR_seq))*100\n",
    "                GC_list.append(GC_content)\n",
    "    mean_GC = np.mean(GC_list)\n",
    "    print(\"La ventana \" + k + \" tiene \" + str(mean_GC) + \" % GC.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "common-bolivia",
   "metadata": {},
   "source": [
    "**El cálculo del % GC medio es correcto para las ventanas seleccionadas.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "incoming-township",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "La ventana Chr13_17000001-18000000 tiene 5.736526750546687 % N.\n",
      "La ventana Chr13_150000001-151000000 tiene 0.2519120160289101 % N.\n",
      "La ventana Chr02_142000001-143000000 tiene 5.477867335809547 % N.\n",
      "La ventana Chr10_99000001-100000000 tiene 3.299987325502919 % N.\n",
      "La ventana Chr07_1-1000000 tiene 8.0760538780196 % N.\n",
      "La ventana Chr05_138000001-138451958 tiene 21.803010408307518 % N.\n"
     ]
    }
   ],
   "source": [
    "# Comprobación del contenido medio de % N.\n",
    "\n",
    "for k, v in comprobaciones.items():\n",
    "    chr_id = v[0]\n",
    "    start = v[1]\n",
    "    end = v[2]\n",
    "    chr_data = data_complete[data_complete[\"Chr\"] == chr_id].sort_values('Start')\n",
    "    chr_values = chr_data[(chr_data[\"Start\"] >= start) & (chr_data[\"Start\"] < end)]\n",
    "    seq = chr_seq[\"Niben261\" + chr_id]\n",
    "    \n",
    "    N_list = []\n",
    "    count = 0\n",
    "    for index, row in chr_values.iterrows():\n",
    "        if count == 0:\n",
    "            if row[\"Start\"] <= start: pass\n",
    "            else:\n",
    "                IR_seq = seq[start-1:row[\"Start\"]-1].upper()\n",
    "                N_content = (IR_seq.count('N')/len(IR_seq))*100\n",
    "                N_list.append(N_content)\n",
    "            new_end = row[\"End\"]\n",
    "        else:\n",
    "            if row[\"Start\"] > new_end + 1:\n",
    "                IR_seq = seq[new_end:row[\"Start\"]-1].upper()\n",
    "                N_content = (IR_seq.count('N')/len(IR_seq))*100\n",
    "                N_list.append(N_content)\n",
    "            new_end = row[\"End\"]\n",
    "        count += 1\n",
    "        if count == len(chr_values):\n",
    "            if row[\"End\"] < end:\n",
    "                IR_seq = seq[row[\"End\"]:end].upper()\n",
    "                N_content = (IR_seq.count('N')/len(IR_seq))*100\n",
    "                N_list.append(N_content)\n",
    "    mean_N = np.mean(N_list)\n",
    "    print(\"La ventana \" + k + \" tiene \" + str(mean_N) + \" % N.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "wireless-stroke",
   "metadata": {},
   "source": [
    "**El cálculo del % N medio es correcto para las ventanas seleccionadas.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "endangered-basin",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>quintile</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Niben261Chr05g1381005</th>\n",
       "      <td>Chr05</td>\n",
       "      <td>138148875</td>\n",
       "      <td>138153885</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Niben261Chr05g1381006</th>\n",
       "      <td>Chr05</td>\n",
       "      <td>138154993</td>\n",
       "      <td>138161476</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Niben261Chr05g1382004</th>\n",
       "      <td>Chr05</td>\n",
       "      <td>138194697</td>\n",
       "      <td>138200365</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Niben261Chr05g1383003</th>\n",
       "      <td>Chr05</td>\n",
       "      <td>138338765</td>\n",
       "      <td>138340782</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         Chr      Start        End  quintile\n",
       "Niben261Chr05g1381005  Chr05  138148875  138153885       4.0\n",
       "Niben261Chr05g1381006  Chr05  138154993  138161476       2.0\n",
       "Niben261Chr05g1382004  Chr05  138194697  138200365       3.0\n",
       "Niben261Chr05g1383003  Chr05  138338765  138340782       1.0"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr_id = \"Chr05\"\n",
    "start = 138000001\n",
    "end = 138451958\n",
    "\n",
    "chr_data = data_complete[data_complete[\"Chr\"] == chr_id].sort_values('Start')\n",
    "chr_values = chr_data[(chr_data[\"Start\"] >= start) & (chr_data[\"Start\"] < end)]\n",
    "chr_values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "public-cleaner",
   "metadata": {},
   "outputs": [],
   "source": [
    "comprobaciones = {\n",
    "    \"Chr13_17000001-18000000\": (\"Chr13\", 17000001, 18000000), # Mejor ventana.\n",
    "    \"Chr13_150000001-151000000\": (\"Chr13\", 150000001, 151000000), # Peor ventana.\n",
    "    \"Chr02_142000001-143000000\": (\"Chr02\", 142000001, 143000000), # Ventana intermedia.\n",
    "    \"Chr10_99000001-100000000\": (\"Chr10\", 99000001, 100000000), # Ventana intermedia.\n",
    "    \"Chr07_1-1000000\": (\"Chr07\", 1, 1000000), # Primera ventana de un cromosoma.\n",
    "    \"Chr05_138000001-138451958\": (\"Chr05\", 138000001, 138451958), # Suitability = 0. Última ventana de un cromosoma.\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "cutting-distance",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Interval</th>\n",
       "      <th>Suitability</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Window</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Mean Length IR</th>\n",
       "      <th>requisite</th>\n",
       "      <th>N genes assigned to expression category</th>\n",
       "      <th>N genes total</th>\n",
       "      <th>Mean % GC</th>\n",
       "      <th>Mean % N</th>\n",
       "      <th>N IR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1895</th>\n",
       "      <td>1895</td>\n",
       "      <td>17000001-18000000</td>\n",
       "      <td>9</td>\n",
       "      <td>Chr13</td>\n",
       "      <td>18</td>\n",
       "      <td>17000001</td>\n",
       "      <td>18000000</td>\n",
       "      <td>41276.666667</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>20</td>\n",
       "      <td>35.146000</td>\n",
       "      <td>5.736527</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2028</th>\n",
       "      <td>2028</td>\n",
       "      <td>150000001-151000000</td>\n",
       "      <td>-133</td>\n",
       "      <td>Chr13</td>\n",
       "      <td>151</td>\n",
       "      <td>150000001</td>\n",
       "      <td>151000000</td>\n",
       "      <td>13156.323529</td>\n",
       "      <td>1</td>\n",
       "      <td>40</td>\n",
       "      <td>68</td>\n",
       "      <td>43.222241</td>\n",
       "      <td>0.251912</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>329</th>\n",
       "      <td>329</td>\n",
       "      <td>142000001-143000000</td>\n",
       "      <td>-21</td>\n",
       "      <td>Chr02</td>\n",
       "      <td>143</td>\n",
       "      <td>142000001</td>\n",
       "      <td>143000000</td>\n",
       "      <td>64738.800000</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>14</td>\n",
       "      <td>32.489838</td>\n",
       "      <td>5.477867</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1536</th>\n",
       "      <td>1536</td>\n",
       "      <td>99000001-100000000</td>\n",
       "      <td>-4</td>\n",
       "      <td>Chr10</td>\n",
       "      <td>100</td>\n",
       "      <td>99000001</td>\n",
       "      <td>100000000</td>\n",
       "      <td>53170.388889</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>17</td>\n",
       "      <td>34.719107</td>\n",
       "      <td>3.299987</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1009</th>\n",
       "      <td>1009</td>\n",
       "      <td>1-1000000</td>\n",
       "      <td>-46</td>\n",
       "      <td>Chr07</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1000000</td>\n",
       "      <td>23214.324324</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>36</td>\n",
       "      <td>34.037078</td>\n",
       "      <td>8.076054</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>860</td>\n",
       "      <td>138000001-138451958</td>\n",
       "      <td>0</td>\n",
       "      <td>Chr05</td>\n",
       "      <td>139</td>\n",
       "      <td>138000001</td>\n",
       "      <td>138451958</td>\n",
       "      <td>86555.200000</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>29.966651</td>\n",
       "      <td>21.803010</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unnamed: 0             Interval  Suitability    Chr  Window      Start  \\\n",
       "1895        1895    17000001-18000000            9  Chr13      18   17000001   \n",
       "2028        2028  150000001-151000000         -133  Chr13     151  150000001   \n",
       "329          329  142000001-143000000          -21  Chr02     143  142000001   \n",
       "1536        1536   99000001-100000000           -4  Chr10     100   99000001   \n",
       "1009        1009            1-1000000          -46  Chr07       1          1   \n",
       "860          860  138000001-138451958            0  Chr05     139  138000001   \n",
       "\n",
       "            End  Mean Length IR  requisite  \\\n",
       "1895   18000000    41276.666667          1   \n",
       "2028  151000000    13156.323529          1   \n",
       "329   143000000    64738.800000          1   \n",
       "1536  100000000    53170.388889          1   \n",
       "1009    1000000    23214.324324          1   \n",
       "860   138451958    86555.200000          1   \n",
       "\n",
       "      N genes assigned to expression category  N genes total  Mean % GC  \\\n",
       "1895                                       14             20  35.146000   \n",
       "2028                                       40             68  43.222241   \n",
       "329                                         7             14  32.489838   \n",
       "1536                                        8             17  34.719107   \n",
       "1009                                       25             36  34.037078   \n",
       "860                                         4              4  29.966651   \n",
       "\n",
       "       Mean % N  N IR  \n",
       "1895   5.736527    21  \n",
       "2028   0.251912    68  \n",
       "329    5.477867    15  \n",
       "1536   3.299987    18  \n",
       "1009   8.076054    37  \n",
       "860   21.803010     5  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# En findIR_compelte.txt, estas ventanas tienen los siguientes valores:\n",
    "list_temp = []\n",
    "for k, v in comprobaciones.items():\n",
    "    chr_id = v[0]\n",
    "    start = v[1]\n",
    "    end = v[2]\n",
    "    chr_data_temp = data_idoneidad_intergenic[data_idoneidad_intergenic[\"Chr\"] == chr_id].sort_values('Start')\n",
    "    temp = chr_data_temp[(chr_data_temp[\"Start\"] >= start) & (chr_data_temp[\"Start\"] < end)]\n",
    "    list_temp.append(temp)\n",
    "data_filter = pd.concat(list_temp)\n",
    "data_filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "thousand-shuttle",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
